vault/custom_response_headers.go

```diff
@@ -0,0 +1,278 @@
+package vault
+
+import (
+	"fmt"
+	log "github.com/hashicorp/go-hclog"
+	"net/http"
+	"net/textproto"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/vault/internalshared/configutil"
+)
+
+type ListenersCustomResponseHeadersList struct {
+	logger log.Logger
+	CustomHeadersList []*ListenerCustomHeaders
+}
+
+type ListenerCustomHeaders struct {
+	Address string
+	StatusCodeHeaderMap map[string][]*CustomHeader
+	// ConfiguredHeadersStatusCodeMap field is introduced so that we would not need to loop through
+	// StatusCodeHeaderMap to see if a header exists, the key for this map is the headers names
+	ConfiguredHeadersStatusCodeMap map[string][]string
+}
+
+type CustomHeader struct {
+	Name string
+	Value string
+}
+
+func NewListenerCustomHeader(ln []*configutil.Listener, logger log.Logger, uiHeaders http.Header) *ListenersCustomResponseHeadersList {
+
+	if ln == nil {
+		return nil
+	}
+
+	ll := &ListenersCustomResponseHeadersList{
+		logger: logger,
+	}
+
+	for _, l := range ln {
+		lc := &ListenerCustomHeaders{
+			Address: l.Address,
+		}
+		lc.StatusCodeHeaderMap = make(map[string][]*CustomHeader)
+		lc.ConfiguredHeadersStatusCodeMap = make(map[string][]string)
+		for sc, hv := range l.CustomResponseHeaders {
+			var chl []*CustomHeader
+			for h, v := range hv {
+				// Sanitizing custom headers
+				// X-Vault- prefix is reserved for Vault internal processes
+				if strings.HasPrefix(h, "X-Vault-") {
+					logger.Error("Custom headers starting with X-Vault are not valid", "header", h)
+					continue
+				}
+
+				// Checking for UI headers, if any common header exists, we just log an error
+				if uiHeaders != nil {
+					exist := uiHeaders.Get(h)
+					if exist != "" {
+						logger.Warn("found a duplicate header in UI, note that config file headers take precedence.", "header:", h)
+					}
```

Comment:
At this time, the decision was made such that configured headers in the config file take precedence over the ones set in the UI endpoint. We have intentionally moved away from having an API endpoint to set these custom headers. So, if an operator would like to update the response headers they would need to update their configuration file and send a SIGHUP signal to the Vault and the custom headers will be updated with the new configured headers. 
There is also an option to modify "/sys/config/reload" endpoint in such a way so that the custom headers would be reloaded as well. As it stands, the endpoint only supports renewing licensing AFAIK. Updating that endpoint for custom response headers may come with a future PR depending on the time and priorities. 
As for deprecating the /ui endpoint, unfortunately, I don't have sufficient information.